index.js

```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
+  return tokens.some(token => ['[', ']', '{}'].includes(token)) ? tokens : []
+}
+
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+
+  // first token can't be '[', ']', or '{}'
+  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
```

Comment:
we checked already in line 108, that the token is only existing once. Maybe it is "cheaper" to store the count once and check if the last token is '{}'

Like this


```js
  let amountOfObjectNotation = 0

  // first token can't be '[', ']', or '{}'
  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
  // only a single {} special ending is allowed
  else if ((amountOfObjectNotation = tokens.filter(t => t === '{}').length) > 1) isValid = false
  // {}, if present, must be the last token
  else if (amountOfObjectNotation === 1 && tokens[tokens.length - 1] !== '{}') isValid = false
```

Not tested, maybe wrong.